### PR TITLE
ppc64le: Restrict maxmem to avoid HTAB allocation failure

### PR DESCRIPTION
**Pull request by nitkon**
Fixes: #363

Signed-off-by: Nitesh Konkar <niteshkonkar@in.ibm.com>

### CODE REVIEW

**Comment by jodh-intel** [comment]
Hi @nitkon - although the commit should pass the `checkcommits` tests, please can you add a sentence or two of description explaining why this change is being made. Also, is 32G an architectural limitation? That's implied but not explicit in this patch.

**Comment by nitkon** [comment]
@jodh-intel :  Yes I shall add a couple of lines in my commit explaining why the limitation on maxmem is needed on PPC64le architecture. However, until that time the detailed explanation could be found in the [here](https://raobharata.wordpress.com/2015/10/09/memory-hotplug-support-in-powerkvm/)

32G is not an architecture limitation. However, assigning large values of maxmem for the kata-container on ppc64le  could result in HTAB failure. Hence I had to choose a memory value that would be (possibly) sufficient for the process inside the kata-container to run and a multiple of 256MB as well.

Your view/suggestions would be appreciated on this.

**Comment by jodh-intel** [comment]
We could consider adding some sort of architecture-specific setting to `configuration.toml` maybe.

**Comment by katabuilder** [comment]
PSS Measurement:
Qemu: 159770 KB
Proxy: 4595 KB
Shim: 10838 KB

Memory inside container:
Total Memory: 2045972 KB
Free Memory:  1995888 KB

**Comment by grahamwhaley** [comment]
Yep, I was thinking we should make it a config toml setting. It might not even have to be architectural - just a cap on the max RAM a container can have in this instance.

**Comment by nitkon** [comment]
@grahamwhaley @jodh-intel : If not architecture specific then what should be the max cap for RAM per container? Any number in mind?

**Comment by grahamwhaley** [comment]
@nitkon - I think the value will be architecture specific, but how you set that value (in a setting in the configuration.toml file) does not have to be - that could be generic.
And then... if there is no setting in the toml, we could default to 'infinite/unlimited'. So, each architecture can then choose if it wants or needs to have a setting in its default toml file.

File: virtcontainers/qemu_ppc64le.go [hunk 1]
```diff
@@ -27,6 +27,8 @@ const defaultQemuMachineOptions = "accel=kvm,usb=off"
 
 const defaultPCBridgeBus = "pci.0"
 
+const defaultMemMaxPPC64le = 32256 // Restrict MemMax to 32Gb on PPC64le
```
[the remark this answers is not in the document]
Comment (nitkon) [reply virtcontainers/qemu_ppc64le.go]: @bergwolf : No, there is no reference for it.  As I mentioned in the above comment ..
```32G is not an architecture limitation. However, assigning large values of maxmem for the kata-container on ppc64le could result in HTAB failure. Hence I had to choose a memory value that would be (possibly) sufficient for the process inside the kata-container to run and a multiple of 256MB as well.```

[the remark this answers is not in the document]
Comment (bergwolf) [reply virtcontainers/qemu_ppc64le.go]: @nitkon per https://bugzilla.redhat.com/show_bug.cgi?id=1282833#c16, the hash table size is 1/128 of the guest maximum memory. Current `32Gb` means that a 100Gb host can at most launch 20 guests (each consuming 256Mb host CMA memory for the hash table) no matter how much memory the guests actually allocate.

How about making the guest maximum memory a parameter instead? Its default value can be host memory size for other arch and `32Gb` for ppc. Also users can set it to whatever size they want. Surely we need to document the client value to help ppc users to choose a right value.

That said, the client option is just an optimization over this PR and can be done in a following up one. So I'm giving it my
LGTM!

Thanks!

[the remark this answers is not in the document]
Comment (WeiZhang555) [reply virtcontainers/qemu_ppc64le.go]: @bergwolf It seems your LGTM can't be counted unless you move it outside.

@nitkon So `32256` is the maximum allowed value on ppc64le but `32512` or bigger number isn't?

**Pull request merged**